cli/src/pcluster/templates/pcluster_dashboard_builder.py

```diff
@@ -0,0 +1,586 @@
+
+        # Head Node logs, if CW Logs are enabled
+        if self.cw_logging_enabled:
+            self._add_cw_log(cloudwatch_dashboard)
+
+    def _update_coord(self, coord, dx, dy):
+        """Calculate coordinates for the new graph. It updates both x and y values if the graph doesn't fit in the
+        given line."""
+        coord.update({"x": coord["x"] + dx})
+        if coord["x"] > 24:  # the longest width allowed is 24
+            coord.update({"x": 0, "y": coord["y"] + dy})
+
+    def _update_coord_after_section(self, coord, dy):
+        """Calculate section header coordinates."""
+        coord.update({"y": coord["y"] + dy})
+
+    def _reset_coord(self, coord):
+        coord.update({"x": 0, "y": 0})
+
+    def _is_logs_section_empty(self, empty_section, section_widgets):
+        """Check if a section will be empty."""
+        empty_section.update({"bool": True})
+        for log_params in section_widgets["widgets"]:
+            if "conditions" in log_params:
+                for cond_dict in log_params["conditions"]:
+                    if cond_dict["param"] in cond_dict["allowed_values"]:
+                        empty_section.update({"bool": False})
+
+    def _add_text_widget(self, markdown, cloudwatch_dashboard):
+        """Add the textwidget to the cloudwatch dashboard and update coordinates."""
+        text_widget = TextWidget(markdown=markdown, height=1, width=24)
+        text_widget.position(x=self.coord["x"], y=self.coord["y"])
+        cloudwatch_dashboard.add_widgets(text_widget)
+        self._update_coord_after_section(self.coord, 1)
+
+    def _generate_graph_widget(self, title, metric_list):
+        graph_width = 6
+        graph_height = 6
+        widget = GraphWidget(
+            title=title,
+            left=metric_list,
+            region=self._stack_region,
+            width=graph_width,
+            height=graph_width,
+        )
+        widget.position(x=self.coord["x"], y=self.coord["y"])
+        self._update_coord(self.coord, graph_width, graph_height)
+        return widget
+
+    def _generate_ec2_metrics_list(self, metrics):
+        metric_list = []
+        for metric in metrics:
+            cloudwatch_metric = cloudwatch.Metric(
+                namespace="AWS/EC2", metric_name=metric, dimensions={"InstanceId": self.head_node_instance.ref}
+            )
+            metric_list.append(cloudwatch_metric)
+        return metric_list
+
+    def _add_conditional_storage_widgets(
+        self,
+        conditional_metrics,
+        volumes_list,
+        namespace,
+        dimension_vol_name,
+        attribute_name,
+        vol_attribute_name,
+    ):
+        """Add widgets for conditional metrics for EBS, Raid and EFS."""
+        widgets_list = []
+        for metric_condition_params in conditional_metrics:
+
+            # Check if there are volumes support the metric
+            is_supported_vol_present = False
+            for volume in volumes_list:
```

Comment:
You can avoid this first loop and do a single loop on which you add the metric if the volume is a supported one.

Reply:
Added a check in the second loops to check if len(metric_list)>0 to aoid the first loop